### DISCUSSION TITLE
Long running Queue list in rabbitmq GUI since migrating to Quorum queues in 3.11.x

### DISCUSSION THREAD

**Discussion by ds2105** [rabbitmq/rabbitmq-server#9817] [Questions] [answered]
Starting with 3.11.1 we switched most of our queues from Classic to Quorum. The total number of queues and the ha-params configuration are almost the same. Its a three node cluster. We noticed that the time to list the queues in rabbitmq GUI did much increases. It takes round about a minute to get a result of queues lists. The number of queues we use are between 15 K up to 25 K. This effect doesn' t changes with any rabbitmq updates even there are made any performance benefits for Quorum queues. The actual version we are using is 3.11.17.

**Comment by michaelklishin** [comment, marked as answer]
This sounds like a duplicate of https://github.com/rabbitmq/rabbitmq-server/discussions/9518.

Are there any currently down cluster peers or former cluster members that could have been removed?

**Reply by ds2105** [reply]
No, it doesn't looks like duplicate to  https://github.com/rabbitmq/rabbitmq-server/discussions/9518. There is no timeout for queues on one of the cluster members. The whole output takes a long time and is much longer than it was for classic queues.

**Reply by ds2105** [reply]
one example:
 rabbitmq-queues quorum_status --vhost PROD  WK.CUPPS.AMADEUS.AMM.A001.DLH.ACUSUSER.AMMM0CKA0002.7.1.FL.A.AMM.T.X.CKI.A0002
Status of quorum queue WK.CUPPS.AMADEUS.AMM.A001.DLH.ACUSUSER.AMMM0CKA0002.7.1.FL.A.AMM.T.X.CKI.A0002 on node rabbit@10-244-108-93 ...

![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/90f3c333-ca69-4728-bb80-3d0c679b36d6)

rabbitmq-queues check_if_node_is_quorum_critical
Checking if node rabbit@10-244-108-91 is critical for quorum of any queues/streams ...
Node rabbit@10-244-108-91 reported no queues/streams with minimum quorum

rabbitmq-queues check_if_node_is_quorum_critical
Checking if node rabbit@10-244-108-92 is critical for quorum of any queues/streams ...
Node rabbit@10-244-108-92 reported no queues/streams with minimum quorum

rabbitmq-queues check_if_node_is_quorum_critical
Checking if node rabbit@10-244-108-93 is critical for quorum of any queues/streams ...
Node rabbit@10-244-108-93 reported no queues/streams with minimum quorum

**Comment by kjnilsson** [comment]
With that many queues I could well imagine it getting a bit slow compared to classic queues. What made you use 15k Quorum queues?

The code I suspect is responsible is `rabbit_quorum_queue:i/2` which ends up doing a few more calls and rpcs than is strictly necessary and could well be optimised. A profiling capture using perf would help us narrow down and possible optimise the right bit of code faster.

**Reply by kjnilsson** [reply]
@ds2105 could you capture the exact HTTP request that takes a long time in the mgmt ui? Things like sorting can affect performance a lot.

I did some local testing with 5k QQs - and removing some potentially slowish work only saved me about 25%, which may of course end up higher in a real environment but still not a huge matter. Also getting the first page of the queue list only took around 1s in the first place. Hence more details on reproduction would be helpful.

**Reply by ds2105** [reply]
Hi Karl, let me answer your second question first. I did some screen dumps with turned on FF Analyzer:
HTTP request in Browser is: https://cloud.cute.dlh.de/rabbitmq/#/queues. This leads to 174 pages with queues, 8 of them are classic queue pages, the rest is quorum.
![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/6dd9891f-6d74-48ad-b95f-07672be22c6a)

Inside the HTTP request FF analyzes this:
![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/db1655cb-5f15-4030-81aa-993fee25544d)

![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/298a7e29-e110-4ba6-875c-2e94bd6a4b30)

![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/8fec3938-dedd-4355-bb26-a9d2a08624ee)

![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/23497935-34a1-41b0-98c0-88fcc3d4d090)

![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/52c23c03-9517-4560-8808-821f6b6c9e08)

![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/c9c1d833-4681-4dd1-9381-8ae9ee373002)

![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/a593b18a-f728-4bdb-bda1-e5dd8930778c)

**Reply by ds2105** [reply]
We are running in a DC/OS cluster with three nodes right now and we have Erlang configured with these additional values:
"RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS": "+S 4:4 +P 524288 +A 64 +MMscs 4096 +sbwt none +sbwtdcpu none +sbwtdio none" 
Each cluster node has 2 CPUs and 16 GB memory.

**Reply by kjnilsson** [reply]
Thanks, so you're running 15k queues on 2 CPU cluster nodes? I suspect you starve the schedulers doing these expensive cluster wide queries. I will do some optimisations around the quorum queue parts which may or may not help. Ultimately for that many queues you are going to need a lot more CPUs to get the parallelism needed to things in a timely manner.

Alternatively you could consider disabling the management metrics: https://www.rabbitmq.com/management.html#disable-stats

**Reply by ds2105** [reply]
Thank you Karl, we will configure the CPU s for each  node higher and give the disabeling for management metrics a try, we are using Prometheus metrics. We also will move from DC/OS to k8s within next weeks switching to a 5 node cluster we tested with. But also with my load tests I made in this k8s environment with 40K QQs, 5 nodes with 8 requested/16 max CPUs, and 20 GB memory for each node I noticed an equal reaction with long responses in Queue monitoring in Rabbitmq GUI. In this test the QQs are there but without having transactions, just idle.

**Reply by michaelklishin** [reply]
There are certain recommendations for clusters that are [mostly idle](https://rabbitmq.com/runtime.html#cpu-reduce-idle-usage) but 15K will never be "completely idle" because they emit stats periodically, even if they do nothing else.

2 cores won't be enough for tens of thousands of queues. Consider using 8 or so.

**Reply by ds2105** [reply]
Thank you Michael, that 's why we will have configured 8 cores initial and with increasing allowed cores until 16 max in k8s environment. That we figured out while mass testing with 40K QQs. The highest resources are used while creating all QQs from scratch. If all QQs are already created the used resources are lower.

**Reply by ds2105** [reply]
By the way - what does a snapshot index "Undefined" mean. I didn't find a explanation for that. Thank you!
![grafik](https://github.com/rabbitmq/rabbitmq-server/assets/53215829/b4e1812e-4bc0-4ecc-8335-0496ff2c6707)

**Reply by michaelklishin** [reply]
A quorum queue may or may not have any log snapshots at the time when these metrics are requested.

**Reply by ds2105** [reply]
OK, thank you. So it isn't an error than a valid value. Understood.

**Reply by michaelklishin** [reply]
We don't have enough information to tell that but very likely it's just the state of the snapshot(s)  (there isn't any)
